source/s3_client.c

```diff
@@ -1926,13 +1905,173 @@ static void s_s3_client_on_host_resolver_address_resolved(
             err_code,
             aws_error_str(err_code));
 
-        s_s3_client_lock_synced_data(client);
+        aws_s3_client_lock_synced_data(client);
         client->synced_data.invalid_endpoint = true;
-        s_s3_client_schedule_process_work_task_synced(client);
-        s_s3_client_unlock_synced_data(client);
+        s_s3_client_schedule_process_work_synced(client);
+        aws_s3_client_unlock_synced_data(client);
     } else {
         AWS_ASSERT(host_addresses);
-        s_s3_client_add_vips(client, host_addresses);
+
+        if (aws_s3_client_add_vips(client, host_addresses)) {
+            int last_error_code = aws_last_error_or_unknown();
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_CLIENT,
+                "id=%p Error %d occurred whileadding VIPs (%s)",
+                (void *)client,
+                last_error_code,
+                aws_error_str(last_error_code));
+        }
+    }
+}
+
+int aws_s3_client_add_vips(struct aws_s3_client *client, const struct aws_array_list *host_addresses) {
+    AWS_PRECONDITION(client);
+    AWS_PRECONDITION(client->vtable);
+    AWS_PRECONDITION(client->vtable->add_vips);
+
+    return client->vtable->add_vips(client, host_addresses);
+}
+
+static int s_s3_client_add_vips_default(struct aws_s3_client *client, const struct aws_array_list *host_addresses) {
+    AWS_PRECONDITION(client);
+    AWS_PRECONDITION(host_addresses);
+
+    struct aws_s3_vip *vip = NULL;
+    int result = AWS_OP_SUCCESS;
+
+    aws_s3_client_lock_synced_data(client);
+
+    if (!client->synced_data.active) {
+        goto unlock;
+    }
+
+    struct aws_byte_cursor server_name = aws_byte_cursor_from_string(client->synced_data.endpoint);
+
+    for (size_t address_index = 0; address_index < aws_array_list_length(host_addresses); ++address_index) {
+
+        if (client->synced_data.active_vip_count >= client->ideal_vip_count) {
+            break;
+        }
+
+        if (client->synced_data.allocated_vip_count >= aws_s3_client_get_max_allocated_vip_count(client)) {
+            AWS_LOGF_WARN(
+                AWS_LS_S3_CLIENT,
+                "id=%p Allocated VIP count (%d) is greater than or equal to twice the amount of allowed VIPs (%d). "
```

Comment:
text here (twice) is kind of couple to the underlying, current threshold, could be reworded to be less likely to be come inaccurate in the future.